react-components/src/architecture/concrete/primitives/tools/PrimitiveEditTool.ts

```diff
@@ -0,0 +1,413 @@
+      } else if (domainObject instanceof PlaneDomainObject) {
+        domainObject.setFocusInteractive(FocusType.None);
+      } else if (domainObject instanceof SolidDomainObject) {
+        domainObject.setFocusInteractive(FocusType.None);
+      }
+    }
+  }
+
+  private initializeStyle(domainObject: DomainObject): void {
+    // Just copy the style the depthTest field from any other selectable
+    const depthTest = this.getDepthTestOnFirstSelectable();
+    if (depthTest === undefined) {
+      return;
+    }
+    const style = domainObject.getRenderStyle();
+    if (!(style instanceof CommonRenderStyle)) {
+      return;
+    }
+    style.depthTest = depthTest;
+  }
+
+  private getDepthTestOnFirstSelectable(): boolean | undefined {
+    for (const otherDomainObject of this.getSelectable()) {
+      const otherStyle = otherDomainObject.getRenderStyle();
+      if (otherStyle instanceof CommonRenderStyle) {
+        return otherStyle.depthTest;
+      }
+    }
+    return undefined;
+  }
+
+  private setDefaultPrimitiveType(): void {
+    if (this.primitiveType === this.defaultPrimitiveType) {
+      return;
+    }
+    this.primitiveType = this.defaultPrimitiveType;
+    CommandsUpdater.update(this.renderTarget);
+  }
+
+  private endCreatorIfFinished(creator: BaseCreator, force = false): void {
+    if (force || creator.isFinished) {
+      this.setDefaultPrimitiveType();
+      this._creator = undefined;
+    }
+  }
+
+  private async tryStartOnHover(event: PointerEvent): Promise<boolean> {
+    const creator = this.createCreator();
+    if (creator === undefined || !creator.canStartOnHover) {
+      return false;
+    }
+    const intersection = await this.getIntersection(event);
+    if (intersection === undefined) {
+      // Click in the "air"
+      return false;
+    }
+    const ray = this.getRay(event);
+    if (!creator.addPoint(ray, intersection, true)) {
+      return false;
+    }
+    const { domainObject } = creator;
+    this.initializeStyle(domainObject);
+    this.deselectAll();
+
+    const parent = this.getOrCreateParent();
+    parent.addChildInteractive(domainObject);
+    domainObject.setSelectedInteractive(true);
+    domainObject.setVisibleInteractive(true, this.renderTarget);
+    this.addTransaction(domainObject.createTransaction(Changes.added));
+    this._creator = creator;
+    return true;
+  }
+
+  // ==================================================
+  // STATIC METHODS
+  // ==================================================
+
+  public static setCursor(
+    tool: BaseTool,
+    domainObject: SolidDomainObject,
+    point: Vector3,
+    pickInfo: PrimitivePickInfo
+  ): void {
+    if (domainObject instanceof BoxDomainObject) {
+      PrimitiveEditTool.setCursorForBox(tool, domainObject, point, pickInfo);
+    } else if (domainObject instanceof CylinderDomainObject) {
+      PrimitiveEditTool.setCursorForCylinder(tool, domainObject, point, pickInfo);
+    }
+  }
+
```

Comment:
I suppose you've already thought about this - but isn't it a bit unfortunate that the `PrimitiveEditTool` must know about the different primitives?

Reply:
You are correct. This is really bad.

Reply:
This made quite a bit of restructuring. I have made a virtual function called getEditToolCursor for getting the cursor while in a edit tool. Also setFocusInteractlve() is made virtual. So now, we don't have to check for specific objects in the PrimitiveEditTool.

Reply:
Also made some matrix transformation simpler by making toViewerMatrix, fromViewerMatrixin the renderTarget. By this, I have removed all CDF_TO_VIEWER_TRANSFORMATION.clone().invert() in the code, since it is stored by the renderTarget instead.